test/common/stats/lazyinit_stats_speed_test.cc

```diff
@@ -0,0 +1,181 @@
+#include "envoy/upstream/upstream.h"
+
+#include "source/common/common/random_generator.h"
+#include "source/common/stats/isolated_store_impl.h"
+#include "source/common/stats/symbol_table.h"
+#include "source/common/stats/thread_local_store.h"
+#include "source/common/thread_local/thread_local_impl.h"
+
+#include "test/benchmark/main.h"
+#include "test/common/stats/real_thread_test_base.h"
+#include "test/test_common/real_threads_test_helper.h"
+
+#include "benchmark/benchmark.h"
+
+namespace Envoy {
+
+namespace Stats {
+
+using Upstream::ClusterTrafficStats;
+
+// Benchmark no-lazy-init on stats, the lazy init version is much faster since no allocation.
+void benchmarkLazyInitCreation(::benchmark::State& state) {
+  const bool lazy_init = state.range(0) == 1;
+  const uint64_t num_stats = state.range(1);
+  Stats::IsolatedStoreImpl stats_store;
+  Upstream::ClusterTrafficStatNames stat_names{stats_store.symbolTable()};
+  std::vector<Stats::ScopeSharedPtr> scopes;
+  std::vector<std::shared_ptr<Stats::LazyInit<ClusterTrafficStats>>> lazy_stats;
+  std::vector<std::shared_ptr<ClusterTrafficStats>> normal_stats;
+
+  for (auto _ : state) { // NOLINT: Silences warning about dead store
+    for (uint64_t i = 0; i < num_stats; ++i) {
+      std::string new_cluster_name = absl::StrCat("cluster_", i);
+      auto scope = stats_store.createScope(new_cluster_name);
+      scopes.push_back(scope);
+      if (lazy_init) {
+        lazy_stats.push_back(
+            std::make_shared<Stats::LazyInit<ClusterTrafficStats>>(*scope, stat_names));
+      } else {
+        normal_stats.push_back(std::make_shared<ClusterTrafficStats>(stat_names, *scope));
+      }
+    }
+  }
+}
+
+BENCHMARK(benchmarkLazyInitCreation)
+    ->ArgsProduct({{0, 1}, {1000, 10000, 20000, 100000}})
+    ->Unit(::benchmark::kMillisecond);
+
+// Benchmark lazy-init of stats in same thread, mimics main thread creation.
+void benchmarkLazyInitCreationInstantiateSameThread(::benchmark::State& state) {
+  const bool lazy_init = state.range(0) == 1;
+  const uint64_t num_stats = state.range(1);
+  Stats::IsolatedStoreImpl stats_store;
+  Upstream::ClusterTrafficStatNames stat_names{stats_store.symbolTable()};
+  std::vector<Stats::ScopeSharedPtr> scopes;
+  std::vector<std::shared_ptr<Stats::LazyInit<ClusterTrafficStats>>> lazy_stats;
+  std::vector<std::shared_ptr<ClusterTrafficStats>> normal_stats;
+
+  for (auto _ : state) { // NOLINT: Silences warning about dead store
+    for (uint64_t i = 0; i < num_stats; ++i) {
+      std::string new_cluster_name = absl::StrCat("cluster_", i);
+      auto scope = stats_store.createScope(new_cluster_name);
+      scopes.push_back(scope);
+      if (lazy_init) {
+        auto lazy_stat = std::make_shared<Stats::LazyInit<ClusterTrafficStats>>(*scope, stat_names);
+        *(*lazy_stat);
+        lazy_stats.push_back(std::move(lazy_stat));
```

Comment:
nit: std::move is not really needed for a shared_ptr.

Reply:
done